doc/design/queue_api.md

```diff
@@ -0,0 +1,178 @@
+# Queue API
+
+@jinzhejz, 10/18/2017
+
+@k82cn, 9/16/2017
+
+## Overview
+
+[Resource sharing architecture for batch and serving workloads in Kubernetes](https://docs.google.com/document/d/1-H2hnZap7gQivcSU-9j4ZrJ8wE_WwcfOkTeAGjzUyLA/edit#) proposed
+`QueueJob` feature to run batch job with services workload in Kuberentes. Considering the complexity, the 
+whole batch job proposal was seperated into two phase: `Queue` and `QueueJob`. This document 
+presents the API deinition of `Queue` for MVP.
+
+### Scope
+
+  * In MVP, the resource request of `QueueJob` are ignored
+  * In MVP, the policy allocates resource based on `Queue`'s configuration
+  * In MVP, `Queue` is namespace level
+
+## Terminology
+
+  * Deserved (Resource): The total number of resources that the arbitrator allocated to the namespace
+  * Overused: The namespace is overused if it used more resources than deserved resources
+  * Underused: The namespace is underused if it used less resource than deserved resources
+
+## API
+
+```go
+type Queue struct {
+    metav1.TypeMeta   `json:",inline"`
+    metav1.ObjectMeta `json:"metadata"`
+
+    Spec QueueSpec
+    Status QueueStatus
+}
+
+type QueueSpec struct {
+    metav1.TypeMeta   `json:",inline"`
+    metav1.ObjectMeta `json:"metadata"`
+
+    // The weight of Queue, which is used by policy to allocatre resource; the 
+    // default value is 1. NOTE: it can not expect allocating more resouce with 
+    // higher weithg, it dependent on policy's reaction to the weight.
+    Weight int
+}
+
+type QueueStatus struct {
+    metav1.TypeMeta   `json:",inline"`
+    metav1.ObjectMeta `json:"metadata"`
+
+    // The deserved resource of Queue according to the policy
+    Deserved    ResourceList
+    // The resources that allocated to Queue, the allocated resource is less or 
+    // equal to `Deserved`:
+    //   * if some resource was preempting, the Allocated is less then Deserved
+    //   * otherwise, Allocated equals to Deserved
+    Allocated   ResourceList
+    // The resource that used by Pod in namespace; if more resource was used than 
+    // Deserved, the overused resource will be preempted.
+    Used        ResourceList
+    // The resources that are preempting for Queue
+    Preempting  ResourceList
+}
+```
+
+## Function Detail
+
+### Workflow
+![workflow](../images/workflow.jpg)
```

Comment:
In workflow.jpg diagram, 
1) I think the queuejob controller might be also added, with same position as the "Quota Manager". It will watch the queue changes as well, but its action should be adjust queuejob size(scale up/down) in the queue changed. 
2) The queuejob objects should also be added into the apiserver, which is the action targets of queuejob controller.

Reply:
got it, could you provide a draft diagram so I can add it into workflow.jpg, or you may update it after this PR is merged.

Reply:
I can send you a draft diagram through email.

Reply:
Update the diagram.

I have a little confuse about the main function of Queue Controller, according to my understanding, it will maintain(add/update/delete) the queue job information(such as priority, resources requirement) in API server. And queue job resources will be calculated/allocated by scheduling policy, right?